### PR TITLE
[jaegermcp] Add get_service_dependencies tool

### PR DESCRIPTION
**Pull request by lopster568**
## Which problem is this PR solving?

- Part of #7827

`QueryService.GetDependencies()` is the only QueryService method with no MCP tool. An agent investigating service relationships has to fetch individual traces and reconstruct the dependency graph manually, even though Jaeger already computes and stores this data.

## Description of the changes

Adds a `get_service_dependencies` tool that returns the service-to-service dependency graph with call counts over a configurable time window (default: last 24h).

**Input:**
- `start_time` (optional): start of time range (RFC3339 or relative like `-24h`). Defaults to `-24h`.
- `end_time` (optional): end of time range (RFC3339 or relative like `now`). Defaults to `now`.

**Output:**
- `dependencies[]`: sorted list of edges, each with `caller`, `callee`, and `call_count`.

**Use cases for an LLM agent:**
- "Which services does payment-service call?" - one tool call
- "What's the blast radius if database-service goes down?" - filter results by callee
- "Show me the service architecture" - full graph in one response

**Files:**
- `internal/handlers/get_dependencies.go` - handler wrapping `QueryService.GetDependencies()`
- `internal/types/get_dependencies.go` - input/output types with jsonschema tags
- `internal/handlers/get_dependencies_test.go` - 10 unit tests (handler + constructor)
- `server.go` - tool registration
- `integration_test.go` - tool discovery + end-to-end invocation test

## How was this change tested?

- `go test ./cmd/jaeger/internal/extension/jaegermcp/...` - all passing
- `make lint` - 0 issues
- `make fmt` - clean

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully: `make lint test`

## AI Usage in this PR (choose one)
- [x] **Light**: AI provided minor assistance (formatting, simple suggestions)

### CODE REVIEW

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	ctx context.Context,
+	_ *mcp.CallToolRequest,
+	input types.GetDependenciesInput,
+) (*mcp.CallToolResult, types.GetDependenciesOutput, error) {
+	lookback := defaultDependencyLookback
+	if input.Lookback != "" {
+		parsed, err := time.ParseDuration(input.Lookback)
+		if err != nil {
+			return nil, types.GetDependenciesOutput{}, fmt.Errorf("invalid lookback: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go]: 2438bd7 - lookback is now validated as positive after parsing. Negative and zero values return a clear error

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	assert.Equal(t, "payment-service", output.Dependencies[0].Child)
+	assert.Equal(t, uint64(150), output.Dependencies[0].CallCount)
+}
+
+func TestGetDependenciesHandler_Handle_WithLookback(t *testing.T) {
+	h := &getDependenciesHandler{queryService: &mockDependencyService{}}
+
+	_, _, err := h.handle(context.Background(), nil, types.GetDependenciesInput{Lookback: "1h"})
+	require.NoError(t, err)
+}
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies_test.go]: 2438bd7 - mock now captures the lookback argument. 
Added separate tests verifying default (24h) and custom (1h) pass through correctly, plus negative and zero edge cases.

File: cmd/jaeger/internal/extension/jaegermcp/server.go [hunk 3]
```diff
@@ -187,6 +187,12 @@ func (s *server) registerTools() {
 			"that determined end-to-end duration. " +
 			"Higher self_time_us values indicate where time is concentrated on the critical path.",
 	}, handlers.NewGetCriticalPathHandler(s.queryAPI))
+
+	mcp.AddTool(s.mcpServer, &mcp.Tool{
+		Name: "get_service_dependencies",
+		Description: "Get the service dependency graph showing which services call which. " +
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegermcp/server.go]:
```suggestion
		Description: "Get the service dependency graph showing caller-callee pairs. " +
```

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2026 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package types
+
+// GetDependenciesInput defines the input parameters for the get_service_dependencies MCP tool.
+type GetDependenciesInput struct {
+	// Lookback defines the time window to query for dependencies (e.g., "1h", "24h").
+	// Defaults to "24h" if not specified.
+	Lookback string `json:"lookback,omitempty" jsonschema:"Time window for dependency data (e.g. 1h, 24h). Default: 24h"`
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go]: this should use start time / end time as SearchTracesInput. The backend API for dependencies requires that, not a lookback.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+// GetDependenciesOutput defines the output of the get_service_dependencies MCP tool.
+type GetDependenciesOutput struct {
+	Dependencies []DependencyLink `json:"dependencies" jsonschema:"Service-to-service dependency edges with call counts"`
+}
+
+// DependencyLink represents a dependency between two services.
+type DependencyLink struct {
+	Parent    string `json:"parent" jsonschema:"Calling service name"`
+	Child     string `json:"child" jsonschema:"Called service name"`
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go]: I would use caller/callee names, no need to introduce another concept to AI which needs explanation (they are not really parent/child)

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	CallCount uint64 `json:"call_count" jsonschema:"Number of calls from parent to child in the time window"`
+	Source    string `json:"source,omitempty" jsonschema:"Source of the dependency data"`
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go]:
```suggestion
```

I am not sure it's useful for AI

**Review by yurishkuro — CHANGES_REQUESTED** [review]
test coverage is incomplete

File: cmd/jaeger/internal/extension/jaegermcp/server.go (hunk 3 above)
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/server.go]: fixed in 58bb3dd

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go (hunk 4 above)
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go]: Switched to start_time / end_time matching the search_traces input pattern. Defaults to last 24h when omitted.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go (hunk 5 above)
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go]: renamed to caller / callee.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go (hunk 6 above)
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go]: Removed. Should have filtered this during initial design rather than exposing all model fields.

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+// GetDependenciesInput defines the input parameters for the get_service_dependencies MCP tool.
+type GetDependenciesInput struct {
+	// StartTime is the start of the time range (optional, defaults to "-24h").
+	// Supports RFC3339 or relative time (e.g., "-1h", "-24h").
+	StartTime string `json:"start_time,omitempty" jsonschema:"Start of time range (RFC3339 or relative like -24h). Default: -24h"`
+
+	// EndTime is the end of the time range (optional, defaults to "now").
+	// Supports RFC3339 or relative time (e.g., "now", "-1h").
+	EndTime string `json:"end_time,omitempty" jsonschema:"End of time range (RFC3339 or relative like now). Default: now"`
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go:14]: done

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+) (*mcp.CallToolResult, types.GetDependenciesOutput, error) {
+	endTime, err := parseDependencyTime(input.EndTime, time.Now())
+	if err != nil {
+		return nil, types.GetDependenciesOutput{}, fmt.Errorf("invalid end_time: %w", err)
+	}
+
+	defaultStart := endTime.Add(-24 * time.Hour)
+	startTime, err := parseDependencyTime(input.StartTime, defaultStart)
+	if err != nil {
+		return nil, types.GetDependenciesOutput{}, fmt.Errorf("invalid start_time: %w", err)
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go:53]: Both default to omitted. A user passing now/now would get a clear error, which is the correct behavior for a zero-duration window.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies_test.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+func TestGetDependenciesHandler_Handle_StartAfterEnd(t *testing.T) {
+	h := &getDependenciesHandler{queryService: &mockDependencyService{}}
+
+	_, _, err := h.handle(context.Background(), nil, types.GetDependenciesInput{
+		StartTime: "now",
+		EndTime:   "-1h",
+	})
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "start_time must be before end_time")
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies_test.go:114]: Covered by StartAfterEnd test. The now/now case is equivalent.

File: cmd/jaeger/internal/extension/jaegermcp/server.go [hunk 10]
```diff
@@ -187,6 +187,12 @@ func (s *server) registerTools() {
 			"that determined end-to-end duration. " +
 			"Higher self_time_us values indicate where time is concentrated on the critical path.",
 	}, handlers.NewGetCriticalPathHandler(s.queryAPI))
+
+	mcp.AddTool(s.mcpServer, &mcp.Tool{
+		Name: "get_service_dependencies",
+		Description: "Get the service dependency graph showing caller-callee pairs. " +
+			"Returns edges with call counts over a configurable time window (default: last 24h).",
+	}, handlers.NewGetDependenciesHandler(s.queryAPI))
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/server.go:195]: Tool schema already exposes these fields with jsonschema descriptions. Duplicating parameter specs in the description was flagged in previous review (#8314).

File: cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+// GetDependenciesOutput defines the output of the get_service_dependencies MCP tool.
+type GetDependenciesOutput struct {
+	Dependencies []DependencyLink `json:"dependencies" jsonschema:"Service-to-service dependency edges with call counts"`
+}
+
+// DependencyLink represents a dependency between two services.
+type DependencyLink struct {
+	Caller    string `json:"caller" jsonschema:"Calling service name"`
+	Callee    string `json:"callee" jsonschema:"Called service name"`
+	CallCount uint64 `json:"call_count" jsonschema:"Number of calls from caller to callee in the time window"`
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/types/get_dependencies.go:26]: Renamed per maintainer review feedback above.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies_test.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+	_, _, err := h.handle(context.Background(), nil, types.GetDependenciesInput{
+		StartTime: startTime.Format(time.RFC3339),
+		EndTime:   endTime.Format(time.RFC3339),
+	})
+	require.NoError(t, err)
+	assert.InDelta(t, (2 * time.Hour).Seconds(), mock.capturedLookback.Seconds(), 1)
+	assert.WithinDuration(t, endTime, mock.capturedEndTs, time.Second)
+}
+
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies_test.go:89]: Relative times are always anchored to time.Now(), matching the search_traces behavior. This is by design, not a gap.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+	}
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go:54]: Same as above. Consistent with parseTimeParam used by search_traces.

File: cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+	links := make([]types.DependencyLink, 0, len(deps))
+	for _, d := range deps {
+		links = append(links, types.DependencyLink{
+			Caller:    d.Parent,
+			Callee:    d.Child,
+			CallCount: d.CallCount,
+		})
+	}
```
[the remark this answers is not in the document]
Comment (lopster568) [reply cmd/jaeger/internal/extension/jaegermcp/internal/handlers/get_dependencies.go:73]: Added sorting by caller then callee, matching the pattern in get_services and get_span_names.

**Review by yurishkuro — APPROVED** [review]
